modcc/solvers.hpp

```diff
@@ -0,0 +1,84 @@
+#pragma once
+
+// Transform derivative block into AST representing
+// an integration step over the state variables, based on
+// solver method.
+
+#include <string>
+#include <vector>
+
+#include "expression.hpp"
+#include "symdiff.hpp"
+#include "symge.hpp"
+#include "visitor.hpp"
+
+expression_ptr remove_unused_locals(BlockExpression* block);
+
+class DirectSolverVisitor : public BlockRewriterBase {
+public:
+    using BlockRewriterBase::visit;
+
+    DirectSolverVisitor() {}
+    DirectSolverVisitor(scope_ptr enclosing): BlockRewriterBase(enclosing) {}
+
+    virtual void visit(AssignmentExpression *e) override {
+        // No solver method, so declare an error if lhs is a derivative.
+        if(auto deriv = e->lhs()->is_derivative()) {
+            error({"The DERIVATIVE block has a derivative expression"
+                   " but no METHOD was specified in the SOLVE statement",
+                   deriv->location()});
+        }
+    }
+};
+
+class CnexpSolverVisitor : public BlockRewriterBase {
+protected:
+    // list of identifier names appearing in derivatives on lhs
+    std::vector<std::string> dvars;
```

Comment:
Happy to apply `_` suffix to protected members as well as private; we had left the protected case up in the air.